src/irr/asset/CGLILoader.cpp

```diff
@@ -0,0 +1,434 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "CGLILoader.h"
+
+//#ifdef _IRR_COMPILE_WITH_GLI_
+#include "gli/gli.hpp"
+//#else
+//	#error "Need GLI"
+//#endif
+
+namespace irr
+{
+	namespace asset
+	{
+		asset::SAssetBundle CGLILoader::loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
+		{
+			if (!_file)
+				return {};
+
+			SContext ctx(_file->getSize());
+			ctx.file = _file;
+
+			const char* filename = _file->getFileName().c_str();
+			gli::texture texture = gli::load(filename);
+			if (texture.empty())
+			{
+				os::Printer::log("Failed to load texture at path ", filename, ELL_ERROR);
+				return {};
+			}
+			
+		    const gli::gl glVersion(gli::gl::PROFILE_GL33);
+			const GLenum target = glVersion.translate(texture.target());
+			const auto format = getTranslatedGLIFormat(texture, glVersion);
+			const uint32_t texelFormatByteSize = getTexelOrBlockBytesize(format.first);
+			IImage::E_TYPE baseImageType;
+			IImageView<ICPUImage>::E_TYPE imageViewType;
+
+			switch (target)
+			{
+				case gli::TARGET_1D:
+				{
+					baseImageType = IImage::ET_1D;
+					imageViewType = ICPUImageView::ET_1D;
+					break;
+				}
+				case gli::TARGET_1D_ARRAY:
+				{
+					imageViewType = ICPUImageView::ET_1D_ARRAY;
+					break;
+				}
+				case gli::TARGET_2D:
+				{
+					baseImageType = IImage::ET_2D;
+					imageViewType = ICPUImageView::ET_2D;
+					break;
+				}
+				case gli::TARGET_2D_ARRAY:
+				{
+					imageViewType = ICPUImageView::ET_2D_ARRAY;
+					break;
+				}
+				case gli::TARGET_3D:
+				{
+					baseImageType = IImage::ET_3D;
+					imageViewType = ICPUImageView::ET_3D;
+					break;
+				}
+				case gli::TARGET_CUBE:
+				{
+					imageViewType = ICPUImageView::ET_CUBE_MAP;
+					break;
+				}
+				case gli::TARGET_CUBE_ARRAY:
+				{
+					imageViewType = ICPUImageView::ET_CUBE_MAP_ARRAY;
+					break;
+				}
+				case gli::TARGET_COUNT:
+				{
+					imageViewType = ICPUImageView::ET_COUNT;
+					break;
+				}
+				default:
+				{
+					assert(0);
+					break;
+				}
+			}
+
+			std::pair<core::smart_refctd_dynamic_array<ICPUImage>, core::smart_refctd_dynamic_array<ICPUImageView>> bundle
```

Comment:
@devshgraphicsprogramming I think I've got it wrong while reading about copyBuffers
It seems to me that there should be one imageView and images in amount of mipmaps and such images will be filled by copy buffers to finaly create imageView

I need to read it one more time for sure

Reply:
One Image and one ImageView

There will be one copy range struct per mip-map (but it will point to 1 image and 1 buffer)